lib/Dialect/FIRRTL/Transforms/ExtractClasses.cpp

```diff
@@ -0,0 +1,146 @@
+//===- ExtractClasses.cpp - Extract OM classes ----------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the ExtractClasses pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLUtils.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/OM/OMOps.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/IRMapping.h"
+#include "llvm/Support/Parallel.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace circt::firrtl;
+using namespace circt::om;
+
+namespace {
+struct ExtractClassesPass : public ExtractClassesBase<ExtractClassesPass> {
+  void runOnOperation() override;
+
+private:
+  void extractClass(FModuleOp moduleOp);
+};
+} // namespace
+
+/// Helper class to capture details about a property.
+struct Property {
+  size_t index;
+  StringRef name;
+  Type type;
+  Location loc;
+};
+
+/// Potentiall extract an OM class from a FIRRTL module which may contain
+/// properties.
+void ExtractClassesPass::extractClass(FModuleOp moduleOp) {
+  // Map from Values in the FModuleOp to Values in the ClassOp.
+  IRMapping mapping;
+
+  // Remember ports and operations to clean up when done.
+  llvm::BitVector portsToErase(moduleOp.getNumPorts());
+  SmallVector<Operation *> opsToErase;
+
+  // Collect information about input and output properties. Mark property ports
+  // to be erased.
+  SmallVector<Property> inputProperties;
+  SmallVector<Property> outputProperties;
+  for (auto [index, port] : llvm::enumerate(moduleOp.getPorts())) {
+    if (!isa<PropertyType>(port.type))
+      continue;
+
+    portsToErase.set(index);
+
+    if (port.isInput())
+      inputProperties.push_back({index, port.name, port.type, port.loc});
+
+    if (port.isOutput())
+      outputProperties.push_back({index, port.name, port.type, port.loc});
+  }
+
+  // If the FModuleOp has no properties, nothing to do.
+  if (inputProperties.empty() && outputProperties.empty())
+    return;
+
+  OpBuilder builder = OpBuilder::atBlockEnd(getOperation().getBody(0));
+
+  // Collect the parameter names from input properties.
+  SmallVector<StringRef> formalParamNames;
+  for (auto inputProperty : inputProperties)
+    formalParamNames.push_back(inputProperty.name);
+
+  // Construct the ClassOp with the FModuleOp name and parameter names.
+  auto classOp = builder.create<ClassOp>(moduleOp.getLoc(), moduleOp.getName(),
```

Comment:
Are we going to use separate `mlir::ModulepOp` to put OM stuffs until the end? I think it makes sense to define class op with the same name symbols as modules, but probably it also means we never be able to put OM under the same ModuleOp.

Reply:
As of now, the symbols don't conflict, since all the FIRRTL modules are under a FIRRTL circuit. If we ever have issues with the symbol conflicts, I imagine we will just add an OM top-level symbol table, which is a child of mlir::ModuleOp, to give a separate namespace.